### PR TITLE
Merge Tokel specific development efforts by Dimxy

### PR DESCRIPTION
**Pull request by NutellaLicka**
Added:
Many TokensV2 updates, bug fixes, and some new Tokel specific RPCs
AssetsV2 optimisations & bug fixes
Oracles dust fix
New "Tokel Standard" token data format for both NFT's and fixed supply tokens (evalcode 0xf7, CC 247)
New royalty feature to tokens for any trades conducted through assets ask/bid transactions.

### CODE REVIEW

**Review by dimxy — APPROVED** [review]
as I am the dev I approve it

File: src/cc/heir.cpp [hunk 1]
```diff
@@ -647,7 +648,7 @@ template <typename Helper> UniValue _HeirFund(int64_t txfee, int64_t amount, std
 
 		// for initial funding do not allow to sign by non-owner key:
 		// if (hasNotMypubkey) {
-        if (TotalPubkeyNormalInputs(mtx, myPubkey) < amount && TotalPubkeyCCInputs(mtx, myPubkey) < amount)
+        if (TotalPubkeyNormalInputs(NULL, mtx, myPubkey) < amount && TotalPubkeyCCInputs(NULL, mtx, myPubkey) < amount)
```
Comment (jmjatlanta) [inline src/cc/heir.cpp:651]: Using `nulllptr` reduces ambiguity.

File: src/cc/old/CCassetstx_v0.cpp [hunk 2]
```diff
@@ -354,21 +354,21 @@ std::string CreateSell(int64_t txfee,int64_t askamount,uint256 assetid,int64_t p
...
             if (CCchange != 0)
                 // change to single-eval or non-fungible token vout (although for non-fungible token change currently is not possible)
-                mtx.vout.push_back(MakeTokensCC1vout((cpAssets->evalcodeNFT) ? cpAssets->evalcodeNFT : EVAL_TOKENS, CCchange, mypk));	
+                mtx.vout.push_back(MakeTokensCC1vout((cpAssets->evalcodeAdd) ? cpAssets->evalcodeAdd : EVAL_TOKENS, CCchange, mypk));	
 
 			std::vector<CPubKey> voutTokenPubkeys;
 			voutTokenPubkeys.push_back(unspendableAssetsPubkey);   
 
-            return FinalizeCCTx(mask, cpTokens, mtx, mypk, txfee, 
+            return FinalizeCCTx(0, cpTokens, mtx, mypk, txfee,
```
Comment (jmjatlanta) [inline src/cc/old/CCassetstx_v0.cpp:371]: can `mask` (line 320) be removed? Also may be able to be removed elsewhere in this file (lines 508, 585, 682).

File: src/cc/old/CCtokens_v0.cpp [hunk 3]
```diff
@@ -900,7 +900,7 @@ std::string TokenTransfer(int64_t txfee, uint256 tokenid, vscript_t destpubkey,
 			std::vector<CPubKey> voutTokenPubkeys;
 			voutTokenPubkeys.push_back(pubkey2pk(destpubkey));  // dest pubkey for validating vout
 
-			return FinalizeCCTx(mask, cp, mtx, mypk, txfee, EncodeTokenOpRet(tokenid, voutTokenPubkeys, std::make_pair((uint8_t)0, vopretEmpty))); 
+			return FinalizeCCTx(0, cp, mtx, mypk, txfee, EncodeTokenOpRet(tokenid, voutTokenPubkeys, std::make_pair((uint8_t)0, vopretEmpty)));
```
Comment (jmjatlanta) [inline src/cc/old/CCtokens_v0.cpp:903]: Remove `mask` (line 859)

**Review by jmjatlanta — COMMENTED** [review]
All of what I've found are minor issues that are more for clarity, with (almost) zero effect on runtime. On the next pass I will focus on functionality. A couple of notes:
- Comments should be comments. Dead code should be removed. The old code will remain in GitHub's history if needed later.
- Replacing `NULL` with `nullptr` is a good idea. It is for more than just readability.
- It can be difficult to spot problems on long lines. About 118 characters max will show on 1 line in GitHub without horizontal scrolling.

File: src/init.cpp [hunk 4]
```diff
@@ -2111,5 +2112,8 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     // SENDALERT
     threadGroup.create_thread(boost::bind(ThreadSendAlert));
 
+    if (KOMODO_NSPV_FULLNODE)
+        fprintf(stderr,"nLocalServices %llx %d, %d\n",(long long)nLocalServices,GetBoolArg("-addressindex", DEFAULT_ADDRESSINDEX),GetBoolArg("-spentindex", DEFAULT_SPENTINDEX));
```
Comment (jmjatlanta) [inline src/init.cpp:2116]: Perhaps use a logger here (if it is initialized).

File: src/rpc/tokensrpc.cpp [hunk 5]
```diff
@@ -105,13 +105,20 @@ UniValue tokenlist(const UniValue& params, bool fHelp, const CPubKey& remotepk)
 UniValue tokenv2list(const UniValue& params, bool fHelp, const CPubKey& remotepk)
 {
     uint256 tokenid;
-    if (fHelp || params.size() > 0)
```
Comment (jmjatlanta) [inline src/rpc/tokensrpc.cpp:108]: tokenid not used.

**Review by jmjatlanta — CHANGES_REQUESTED** [review]
Most are minor, some are opinions. 1 is a segfault, albeit probably rare.

**Comment by dimxy** [comment]
I have gone through all of those above notes and made improvements/cleanups. Thank you @jmjatlanta 
(ofc there are still much to improve yet)

File: src/cc/CCassetsCore_impl.h [hunk 6]
```diff
@@ -115,16 +116,17 @@ uint8_t SetAssetOrigpubkey(std::vector<uint8_t> &origpubkey_out, CAmount &unit_p
...
     int32_t n; 
     uint8_t vintxFuncId; 
 	std::vector<uint8_t> origpubkey; 
 	CScript script;
 	uint8_t evalCode;
+    int32_t expiryHeight;
 
     n = vintx.vout.size();
-    if( n == 0 || (vintxFuncId = A::DecodeAssetTokenOpRet(vintx.vout[n-1].scriptPubKey, evalCode, assetid, assetid2, price, origpubkey)) == 0 ) {
+    if( n == 0 || (vintxFuncId = A::DecodeAssetTokenOpRet(vintx.vout.back().scriptPubKey, evalCode, assetid, price, origpubkey, expiryHeight)) == 0 ) {
```
Comment (jmjatlanta) [inline src/cc/CCassetsCore_impl.h:129]: If you used vintx.vout.size() == 0 here `n` would no longer be required.

**Review by jmjatlanta — COMMENTED** [review]
The recent changes look good. I rank this one important: 
https://github.com/KomodoPlatform/komodo/pull/467/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=.include&file-filters%5B%5D=.py&file-filters%5B%5D=.sh&file-filters%5B%5D=No+extension#r671461098

I am getting an assertion with komodo-test. It may be my environment, or something that has already been fixed:
```
./src/komodo-test --gtest_filter=TestAddrmanTests.addrman_simple
Note: Google Test filter = TestAddrmanTests.addrman_simple
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from TestAddrmanTests
[ RUN      ] TestAddrmanTests.addrman_simple
komodo-test: addrman.cpp:160: void CAddrMan::ClearNew(int, int): Assertion `infoDelete.nRefCount > 0' failed.
```

**Pull request merged**